src/config_manager.cpp

```diff
@@ -0,0 +1,79 @@
+#include "config_manager.h"
+#include "rosflight.h"
+
+namespace rosflight_firmware
+{
+ConfigManager::ConfigManager(ROSflight &RF, config_t &config):
+  RF_{RF},
+  config_{config}
+{}
+
+bool ConfigManager::init()
+{
+  if(!read())
+    fill_defaults();
+  return true;
+}
+
+bool ConfigManager::configure_devices()
+{
+  bool success = true;
+  for(device_t device{static_cast<device_t>(0)}; device < Configuration::DEVICE_COUNT; ++device)
+    success = success &&RF_.board_.enable_device(device, config_.config[device], RF_.params_);
+  return success;
+}
+
+ConfigManager::config_response ConfigManager::attempt_set_configuration(device_t device, uint8_t config)
+{
+  config_response resp = RF_.board_.get_board_config_manager().check_config_change(device, config);
+  if(resp.successful)
+    set_configuration(device, config);
+  return resp;
+}
+void ConfigManager::set_configuration(device_t device, uint8_t config)
+{
+  config_.config[device] = config;
+  // TODO consider deinitializing and changing the config
+}
+uint8_t ConfigManager::get_configuration(device_t device)
+{
+  return config_.config[device];
+}
+uint8_t ConfigManager::operator[](device_t device)
+{
+  return get_configuration(device);
+}
+
+void ConfigManager::prepare_write()
+{
+  config_.checksum = generate_checksum();
+}
+
+bool ConfigManager::read()
+{
+  if(!RF_.memory_manager_.is_ready())
+    return false;
+  if(generate_checksum() != config_.checksum)
+    return false;
+  return true;
+}
+
+void ConfigManager::fill_defaults()
+{
+  memset(config_.config, 0, sizeof(config_.config));
+}
+uint32_t ConfigManager::generate_checksum()
+{
+  //8 bit fletcher algorithm, because we can't assume that the struct is a multiple of 16 bits
+  const uint8_t *config_data = reinterpret_cast<const uint8_t*>(config_.config);
+  uint8_t check_a{0};
+  uint8_t check_b{0};
+  for(size_t index{0}; index< sizeof(config_.config); index++ )
+  {
+    check_a += config_data[index];
+    check_b += check_a;
+  }
+  return check_a & (check_b<<8) & (~check_a<<16) & (~check_b<<24);
+}
```

Comment:
I think this is not a true Fletcher checksum, because it's missing the modulo 255 step on `check_a` and `check_b`. You'll also want to make sure that `check_a` and `check_b` don't overflow if you're doing an actual Fletcher checksum, which means they should probably be `uint16_t`, and there should either be a period application of the modulo or a `static_assert` that `sizeof(config_.config)` isn't big enough to cause an overflow.

Fixing this up could probably be left to a separate PR request though, because we should probably transition the parameters and backup memory to use a Fletcher checksum as well, which should all share a common implementation.